### DISCUSSION TITLE
Bluetooth driver on Ubuntu install

### DISCUSSION THREAD

**Discussion by nborof** [thorrak/fermentrack#632] [Q&A] [answered]
I have installed Fermentrack on a laptop running Ubuntu 20.04.  Everything works, except the bluetooth dongle I had isn't strong enough to reliably read the tilt hydrometer in my fermentation chamber.  I found a new (stronger) bluetooth dongle, which required downloading a driver.  It works fine in Ubuntu, but Fermentrack doesn't seem to get any results, even when I place it on right top of the fermentor.   This leads me to believe Fermentrack doesn't have access to the new driver.

Is there a way to get Fermentrack (or docker) to run with an added bluetooth driver on Ubuntu?

**Comment by thorrak** [comment, marked as answer]
What dongle (and what driver)?

**Reply by nborof** [reply]
It is the ZEXMTE Long Range Bluetooth 5.0 adapter with the RTL8761B chipset.  I downloaded and installed the driver from https://raw.githubusercontent.com/Realtek-OpenSource/android_hardware_realtek/rtk1395/bt/rtkbt/Firmware/BT/rtl8761b_fw

**Reply by nborof** [reply]
Here's an example of one: https://www.amazon.com/gp/product/B08LVH5BCP

**Reply by thorrak** [reply]
I'm drilling through a different problem at the moment, but we've got a lot of road left on this trip. I'll try to take a look at the docs for that driver next and see if there is an easy way to tweak the docker compose file to incorporate.

**Reply by thorrak** [reply]
What all did you have to do in order to install the drivers? Do you have a link to instructions?

**Reply by nborof** [reply]
I just copied the “rtl8761b_fw” file to  /lib/firmware/rtl_bt/ and renamed it to “rtl8761b_fw.bin” and rebooted

**Reply by nborof** [reply]
If it's any help, here's another location of a version of the driver.  It's a different file size from the other driver I linked to, but I can confirm it works. https://aur.archlinux.org/packages/rtl8761b-fw/

**Reply by nborof** [reply]
I feel a bit embarrassed, but somehow, the bluetooth dongle seems to be working now!  I don't know what (if anything) changed.  I spent a little time reinstalling Fermentrack and playing around with Docker today, but I can't point to anything I did that would have solved the problem.  If I unplug the dongle and plug it back in, it doesn't work with Fermentrack until I restart the computer (it works plug-and-play with everything else in Ubuntu, however). So I guess the problem is solved? Maybe I made it up in my head? : /

Thank you so much for taking the time to look into this issue.  And thank you for taking the time to write this software in the first place! I'm looking forward to using it over the next few days.

**Reply by nborof** [reply]
I suppose it’s possible that when I reinstalled Fermentrack, something (Docker?) detected the Bluetooth driver (which wasn’t on the computer when I first installed Fermentrack).

**Reply by thorrak** [reply]
What you experienced makes sense - it sounds like the drivers need to be loaded in the base OS before Docker loads the Fermentrack container. Restarting the Pi accomplishes this - as does reinstalling Fermentrack.

If you have any further issues, you may be able to edit Fermentrack's docker-compose.yml file to map your driver into the container, but I would avoid this unless necessary. No reason to modify things if they work without change!

**Reply by nborof** [reply]
If the usb dongle is unplugged for any reason (or if the computer goes to sleep), Fermentrack stops recognizing the dongle until the computer restarts, but I also found that running "docker-compose down; docker-compose up -d" from the "fermentrack-tools" directory gets it working again.  I'm posting that just so anyone else in a similar situation will see. 

I agree - no need to modify things if they're working!

**Comment by lbussy** [comment]
> Everything works, except the bluetooth dongle I had isn't strong enough to reliably read the tilt hydrometer in my fermentation chamber.

I know you've already bought the BLE dongle, but have you looked at [TiltBridge](https://www.tiltbridge.com/) as a potentially better solution?  It's a little less expensive (not sure if you can return what you have) and I assume easier to get working than what you are going through now.

There's are also some additional capabilities you may enjoy.

**Reply by nborof** [reply]
I looked at TiltBridge, but it seems like the hardware is either unavailable (the AliExpress link on the TiltBridge website says board and the screen are no longer available) or more expensive than the $18 I paid for the dongle.  I do have an old phone I can put on the side of the fridge, so I can just use that (along with the Tilt app) for now.  I really like Fermentrack's interface (and obviously its capabilities), so I was just looking to see if there was a little line of code I could tweak and have it work.  I've read there are ways to run docker with access to specific devices on the host computer, but I don't know if that would work here.

Regardless, I can certainly get by with the old phone and Google Sheets for now.  I just thought I'd ask to see if there was an easy fix.  Thanks for your responses. :-)

**Reply by lbussy** [reply]
[One on Amazon for $16.00.](https://smile.amazon.com/ICQUANZX-T-Display-Bluetooth-Development-Arduino/dp/B07VNG9D52) And I guess it depends on what your time is worth to you.  It takes < 15 minutes to get TiltBridge running.  Not sure how much time you have in this so far. :)

And believe me, I've spent more time trying to "Save money" than I care to count.  I speak from shared pain, not from a position of judgement. :)

**Reply by nborof** [reply]
Thanks for the link! Yes, I think my “spend as little as possible” mindset has been overactive since shelling out $140 for the Tilt.

**Reply by nborof** [reply]
lbussy - I ended up buying one of the TTGO ESP32s that you linked to, but whenever I try to connect via USB to flash it, I get 

Connecting........_____....._____....._____....._____....._____....._____....._____
/dev/cu.Bluetooth-Incoming-Port failed to connect: Failed to connect to ESP32: Timed out waiting for packet header

I tried multiple cables, holding down different combinations of buttons, and nothing has worked.  Alas, <15 minutes is now pushing a couple hours...

Any advice?

![Screen Shot 2021-05-13 at 7 48 30 PM](https://user-images.githubusercontent.com/38414932/118204603-37a03280-b424-11eb-8558-94be5900617d.png)

**Reply by nborof** [reply]
Never mind.  I got it to work through Fermentrack.  :-)

**Reply by thorrak** [reply]
If you have any issues whatsoever, let me know. TiltBridge was made for Fermentrack, after all. ;)

**Reply by thorrak** [reply]
> lbussy - I ended up buying one of the TTGO ESP32s that you linked to, but whenever I try to connect via USB to flash it, I get
> 
> Connecting........_****.....****_....._****.....****_....._****.....****_....._____
> /dev/cu.Bluetooth-Incoming-Port failed to connect: Failed to connect to ESP32: Timed out waiting for packet header
> 
> I tried multiple cables, holding down different combinations of buttons, and nothing has worked. Alas, <15 minutes is now pushing a couple hours...
> 
> Any advice?
> 
> ![Screen Shot 2021-05-13 at 7 48 30 PM](https://user-images.githubusercontent.com/38414932/118204603-37a03280-b424-11eb-8558-94be5900617d.png)

For what it's worth (and in case anyone else sees this!) Macs require a driver to be able to recognize the serial-to-USB chips on some ESP32 (and ESP8266) boards. That driver is included in the linux distribution that most Raspberry Pis run, hence why they work with Fermentrack but not on Macs out of the box.
